### DISCUSSION TITLE
RFC: fully featured language server

### DISCUSSION THREAD

**Discussion by lukepighetti** [mint-lang/mint#270] [Language Development]
Hi there, I'm not sure if this already exists, or if it's in the pipeline, and what stage it's in... but I'm used to working with Dart which has an extremely robust language server that provides click-through support to definitions and also type information with inline documentation on hover. It also provides instant feedback as an IDE error with highlighting if you are trying to access a class member that doesn't exist. Here's a quick gif showing these features.

It would be wonderful if Mint would offer a similar feature set and if we could prioritize a VSCode plugin to handle this functionality.

![ezgif-2-35c035348d34](https://user-images.githubusercontent.com/666539/80894869-400c4c00-8cad-11ea-98ef-63c14ced701a.gif)

<img width="837" alt="Screen Shot 2020-05-02 at 7 46 53 PM" src="https://user-images.githubusercontent.com/666539/80894927-b0b36880-8cad-11ea-8c16-020cf8838d35.png">

**Comment by gdotdesign** [comment]
I've started working on this a while back in #203 also worked on it a little bit. 

It's a long task but I think it can be shipped in parts. The first one probably will be the hover info.

**Comment by lukepighetti** [comment]
@gdotdesign do we have an eta for a most basic language server? Undefined getter/setter/method, class members auto-complete, and click to definition would be a wonderful starting point... but I agree, any feature at all would be good to get tooling (such as https://github.com/s0kil/mint-vscode) setup and aligned for future features

**Comment by gdotdesign** [comment]
Currently my focus is mint-ui (UI library) and the language server, depending on which I have more motivation at the time.

I think the hover provider can be shipped in the first release, which is around 50% complete, so probably in a few weeks (next release). After that we need to figure out which feature of it should get more focus.

**Comment by lukepighetti** [comment]
Please ping me when you have a minimal language server that can be incorporated into `mint-vscode`. It's not important to me how fully featured it is at this point.

**Comment by gdotdesign** [comment]
_Tracking progress on the features here: https://github.com/mint-lang/mint/blob/language-server/src/ls/README.md_
_The implementation of the protocol is here: https://github.com/mint-lang/mint/blob/language-server/src/lsp/README.md_

So started working on this again and cleaned up and extended what was there. 

The following features are working but very basic still:

### Completion

A lot of things are available for completion, there are three parts of it:

- **scoped based completion** - for example arguments if the cursor is inside a functions, or functions, styles and properties if the cursor is inside a component
- **global completion** -  functions / constants of stores, modules and global components
- **components** - HTML snippets for all components

![completion](https://user-images.githubusercontent.com/302692/104936960-84f82000-59ad-11eb-8b33-9e5bacdab513.gif)

### Formatting

The formatting command and format on save works (only if the file is valid AST wise):

![auto-save](https://user-images.githubusercontent.com/302692/104936966-86c1e380-59ad-11eb-8f53-8aa404907652.gif)

### Type information on hover

The type of any node can be seen on hover:

![hover](https://user-images.githubusercontent.com/302692/104938324-364b8580-59af-11eb-83d6-50a06221481f.gif)

### Next Steps

For the next steps will be polishing these features a bit more and then releasing it for the next version. After that we can start implementing new features.

### How do I test this?

You can test this by compiling from source and setting it up in your editor: basically the command to start a language server is `mint ls`

I've only tested it in Sublime Text and for that this is the configuration:
```
{
  "mint": {
    "command": ["/path/to/binary/mint", "ls"],
    "scopes": ["source.mint"],
    "syntaxes": ["Packages/Mint/mint.sublime-syntax"],
    "languageId": "mint",
    "enabled": true
  }
}
```

And here is the syntax file:
[mint.sublime-syntax.zip](https://github.com/mint-lang/mint/files/5831288/mint.sublime-syntax.zip)

### What can you do to move this along?

- try it out with your editor and give feedback how it works / behaves
- gather information about how these features could be improved
- If you feel like it implement some missing functionality

-----------------

@lukepighetti @s0kil @Namek  I'm pinging you because you were interested in this topic.

**Comment by gdotdesign** [comment]
I've published a new version of the VSCode extension which supports the language server. 

Once installed you need to set the path to the binary in the settings for it to work (and maybe restart it):
![mint-vscode-settings](https://user-images.githubusercontent.com/302692/105185862-573be400-5b31-11eb-8f9a-57020115ae84.png)

**Reply by lukepighetti** [reply]
Really exciting, thanks for the update.

**Comment by scroobius-pip** [comment]
Thanks for the update
